packages/react/src/file-upload/FileUpload.tsx

```diff
@@ -102,28 +104,105 @@ export const FileUpload = forwardRef<HTMLInputElement, FileUploadProps>(
 		const maxSizeBytes = maxSize && !isNaN(maxSize) ? maxSize * 1000 : 0;
 		const formattedMaxFileSize = formatFileSize(maxSizeBytes);
 
-		const [fileRejections, setFileRejections] = useState<RejectedFile[]>([]);
+		if (maxFiles !== undefined && maxFiles < 1) {
+			throw new Error('maxFiles cannot be less than 1');
+		}
+
+		const [acceptedFiles, setAcceptedFiles] = useState<FileWithStatus[]>(value);
+		const [tooManyFilesRejections, setTooManyFilesRejections] = useState<
+			FileRejection[]
+		>([]);
+		const [invalidRejections, setInvalidRejections] = useState<FileRejection[]>(
+			[]
+		);
+
+		const allFiles = [
+			...acceptedFiles,
+			...tooManyFilesRejections.map((rej) => rej.file),
+			...invalidRejections.map((rej) => rej.file),
+		];
+
+		const allRejections = [...invalidRejections, ...tooManyFilesRejections];
+
+		const changeToken = JSON.stringify(acceptedFiles);
+		useEffect(() => {
+			onChange?.(acceptedFiles);
+			// eslint-disable-next-line react-hooks/exhaustive-deps
+		}, [changeToken]);
+
+		const handleRemoveAcceptedFile = (index: number) => {
+			clearErrors();
+			setAcceptedFiles((prevAcceptedFiles) => {
+				const updatedAcceptedFiles = removeItemAtIndex(
+					prevAcceptedFiles,
+					index
+				);
+				if (tooManyFilesRejections.length) {
+					const newAcceptedFile = tooManyFilesRejections[0].file;
+					handleRemoveWaitingListItem(0, newAcceptedFile);
 
-		const handleRemoveFile = (file: FileWithStatus) => {
-			const indexOfFile = value.indexOf(file);
-			onChange(value.filter((_, index) => index !== indexOfFile));
+					return [...updatedAcceptedFiles, newAcceptedFile];
+				}
+				return updatedAcceptedFiles;
+			});
 		};
 
-		const handleRemoveRejection = (fileName: string) => {
-			setFileRejections(
-				fileRejections.filter((err) => err.file.name !== fileName)
-			);
+		const handleRemoveWaitingListItem = (
+			index: number,
+			file?: FileWithStatus
+		) => {
+			setTooManyFilesRejections((prevWaitingList) => {
+				if (file && prevWaitingList[index]?.file.name === file.name) {
+					return removeItemAtIndex(prevWaitingList, index);
+				} else if (!file) {
+					return removeItemAtIndex(prevWaitingList, index);
```

Comment:
Both code branches do the same thing 🤷  Should we combine them?

Reply:
Maybe flip the logic so that you early return when you want `prevWaitingList` and `removeItemAtIndex(prevWaitingList, index)` all other times?

Reply:
Actually I should be able to just remove this code as this was code written specifically to remove a single "too many files" error which isn't possible any more. I suspect there might be a fair bit of legacy code like this still in the PR.

Reply:
code removed